README.md
# Pathinator

> SVG path builder and minifier
>
> [![npm][npm]][npm-url]
[![build][build]][build-url]
[![coverage][coverage]][coverage-url]
[![deps][deps]][deps-url]
[![size][size]][size-url]
[![vulnerabilities][vulnerabilities]][vulnerabilities-url]
[![license][license]][license-url]


<br><a name="Installation"></a>

## Installation
```
npm install pathinator
```

<br><a name="optimizeSvgPaths"></a>

## optimizeSvgPaths(input, [settings]) ⇒ <code>Promise.&lt;(string\|Document)&gt;</code>
> Optimize all the paths in an svg file. Currently supports the `d` attribute in the `path` element and the `points` attribute in both `polygon` and `polyline` elements.

**Returns**: <code>Promise.&lt;(string\|Document)&gt;</code> - Resolves with the optimized svg in the same format provided.  

| Param | Type | Description |
| --- | --- | --- |
| input | <code>string</code>, <code>Document</code> | Can be the string content of an svg or a DOM Document. |
| [settings] | <code>object</code> | See [Path.export](#Path+export). Settings are applied to each path found in the svg. |

**Example**  
``` javascript
import { optimizeSvgPaths } from 'pathinator';

const svg = document.querySelector('svg');

optimizeSvgPaths(svg, { compress: true })
    .then((result) => {
        // do something
    });
```

<br><a name="Path"></a>

## Path

* [Path](#Path)
    * [new Path([path])](#new_Path_new)
    * [.length](#Path+length) ⇒ <code>integer</code>
    * [.import(path)](#Path+import) ⇒ <code>object</code>
    * [.move(...args)](#Path+move) ⇒ <code>object</code>
    * [.line(...args)](#Path+line) ⇒ <code>object</code>
    * [.cubic(...args)](#Path+cubic) ⇒ <code>object</code>
    * [.quadratic(...args)](#Path+quadratic) ⇒ <code>object</code>
    * [.arc(...args)](#Path+arc) ⇒ <code>object</code>
    * [.close([...args])](#Path+close) ⇒ <code>object</code>
    * [.update(index, values)](#Path+update) ⇒ <code>object</code>
    * [.eachPoint(callback)](#Path+eachPoint) ⇒ <code>object</code>
    * [.transform([settings])](#Path+transform) ⇒ <code>object</code>
    * [.export([settings])](#Path+export) ⇒ <code>Promise.&lt;string&gt;</code>


<br><a name="new_Path_new"></a>

### new Path([path])
> Parse, build, and optimize SVG path data.


| Param | Type | Description |
| --- | --- | --- |
| [path] | <code>string</code> | Optional path data to parse. |

**Example**  
``` javascript
import { Path } from 'pathinator';

const path = new Path()
    .move(50, 100)
    .line(100, 100)
    .line(200, 200)
    .close();
```

<br><a name="Path+length"></a>

### path.length ⇒ <code>integer</code>
> The total number of commands in this path.


<br><a name="Path+import"></a>

### path.import(path) ⇒ <code>object</code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;_`🔗 Chainable`_

> Import a path string. Removes any previous commands and create a new one.

**Returns**: <code>object</code> - Returns this.  

| Param | Type | Description |
| --- | --- | --- |
| path | <code>string</code> | A valid path data string or polygon string. |


<br><a name="Path+move"></a>

### path.move(...args) ⇒ <code>object</code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;_`🔗 Chainable`_

> Add a [move](https://www.w3.org/TR/SVG/paths.html#PathDataMovetoCommands) command.

**Returns**: <code>object</code> - Returns this.  

| Param | Type | Description |
| --- | --- | --- |
| ...args | <code>\*</code> | X and y coordinates or a string of X and y coordinates. If the final argument is `true` then command will be absolute coordinates. |


<br><a name="Path+line"></a>

### path.line(...args) ⇒ <code>object</code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;_`🔗 Chainable`_

> Add a [line](https://www.w3.org/TR/SVG/paths.html#PathDataLinetoCommands) command.

**Returns**: <code>object</code> - Returns this.  

| Param | Type | Description |
| --- | --- | --- |
| ...args | <code>\*</code> | X and y coordinates or a string of X and y coordinates. If the final argument is `true` then command will be absolute coordinates. |


<br><a name="Path+cubic"></a>

### path.cubic(...args) ⇒ <code>object</code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;_`🔗 Chainable`_

> Add a [quadratic bezier curve](https://www.w3.org/TR/SVG/paths.html#PathDataCubicBezierCommands) command.

**Returns**: <code>object</code> - Returns this.  

| Param | Type | Description |
| --- | --- | --- |
| ...args | <code>\*</code> | Series of coordinates or a string of coordinates. If the final argument is `true` then command will be absolute coordinates. |


<br><a name="Path+quadratic"></a>

### path.quadratic(...args) ⇒ <code>object</code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;_`🔗 Chainable`_

> Add a [quadratic bezier curve](https://www.w3.org/TR/SVG/paths.html#PathDataQuadraticBezierCommands) command.

**Returns**: <code>object</code> - Returns this.  

| Param | Type | Description |
| --- | --- | --- |
| ...args | <code>\*</code> | Series of coordinates or a string of coordinates. If the final argument is `true` then command will be absolute coordinates. |


<br><a name="Path+arc"></a>

### path.arc(...args) ⇒ <code>object</code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;_`🔗 Chainable`_

> Add an [arc](https://www.w3.org/TR/SVG/paths.html#PathDataEllipticalArcCommands) command.

**Returns**: <code>object</code> - Returns this.  

| Param | Type | Description |
| --- | --- | --- |
| ...args | <code>\*</code> | Series of coordinates / values or a string of coordinates / values. If the final argument is `true` then command will be absolute coordinates. |


<br><a name="Path+close"></a>

### path.close([...args]) ⇒ <code>object</code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;_`🔗 Chainable`_

> Add a [close](https://www.w3.org/TR/SVG/paths.html#PathDataClosePathCommand) command.

**Returns**: <code>object</code> - Returns this.  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| [...args] | <code>boolean</code> | <code>false</code> | If the argument is `true` then command will be absolute coordinates. |


<br><a name="Path+update"></a>

### path.update(index, values) ⇒ <code>object</code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;_`🔗 Chainable`_

> Update command values at a specific index.

**Returns**: <code>object</code> - Returns this.  

| Param | Type | Description |
| --- | --- | --- |
| index | <code>integer</code> | Index of the command to update. |
| values | <code>string</code>, <code>Array.&lt;number&gt;</code> | New values for the command at this index. |


<br><a name="Path+eachPoint"></a>

### path.eachPoint(callback) ⇒ <code>object</code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;_`🔗 Chainable`_

> Calls  callback for each point in the path.

**Returns**: <code>object</code> - Returns this.  

| Param | Type | Description |
| --- | --- | --- |
| callback | <code>function</code> | Provides three arguments: the Point, a boolean indicating if the point is a control point, and the command index. |


<br><a name="Path+transform"></a>

### path.transform([settings]) ⇒ <code>object</code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;_`🔗 Chainable`_

> Transform all commands in path.

**Returns**: <code>object</code> - Returns this.  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| [settings] | <code>object</code> |  | Optional settings object. |
| [settings.fractionDigits] | <code>integer</code> | <code>3</code> | Round all numbers in path to a specified number of fraction digits. |
| [settings.scale] | <code>number</code>, <code>Point</code>, <code>Array</code>, <code>object</code> |  | Scale the entire path. If a number is provided then x and y are scaled the same. To scale x and y differently provide a Point, an array as [x, y], or an object as { x:_, y:_ }. |
| [settings.translate] | <code>number</code>, <code>Point</code>, <code>Array</code>, <code>object</code> |  | Translate the entire string a specified distance. If a number is provided then x and y are translated the same. To translated x and y differently provide a Point, an array as [x, y], or an object as { x:_, y:_ }. |


<br><a name="Path+export"></a>

### path.export([settings]) ⇒ <code>Promise.&lt;string&gt;</code>
> Export a string of the path. Transforms are only applied to the exported path.


| Param | Type | Default | Description |
| --- | --- | --- | --- |
| [settings] | <code>object</code> |  | Optional settings object. |
| [settings.coordinates] | <code>string</code> | <code>&quot;initial&quot;</code> | Can be 'absolute' to convert all coordinates to absolute, 'relative' to convert all coordinates to relative, 'auto' to convert coordinates to whichever is the fewest characters, 'initial' (default) to retain the coordinates set on each command. |
| [settings.compress] | <code>boolean</code> | <code>false</code> | Remove excess whitespace and unnecessary characters. |
| [settings.combine] | <code>boolean</code> | <code>true</code> | Combine consecutive commands that are redundant. |
| [settings.fractionDigits] | <code>integer</code> | <code>3</code> | Round all numbers in path to a specified number of fraction digits. |
| [settings.scale] | <code>number</code>, <code>Point</code>, <code>Array</code>, <code>object</code> |  | Scale the entire path. If a number is provided then x and y are scaled the same. To scale x and y differently provide a Point, an array as [x, y], or an object as { x:_, y:_ }. |
| [settings.translate] | <code>number</code>, <code>Point</code>, <code>Array</code>, <code>object</code> |  | Translate the entire string a specified distance. If a number is provided then x and y are translated the same. To translated x and y differently provide a Point, an array as [x, y], or an object as { x:_, y:_ }. |
| [settings.maxCharsPerLine] | <code>integer</code> |  | Add newlines at logical breaks in the path to improve readability. |
| [settings.commandsOnNewLines] | <code>boolean</code> | <code>false</code> | Add a newline between each command. |
| [settings.toPolygon] | <code>boolean</code> | <code>false</code> | Format the string for use in a polygon element. Sets coordinates to 'absolute'. |
| [settings.async] | <code>boolean</code> | <code>false</code> | Process each command in a separate Promise. |


[npm]: https://img.shields.io/npm/v/pathinator.svg
[npm-url]: https://npmjs.com/package/pathinator
[build]: https://travis-ci.org/DarrenPaulWright/pathinator.svg?branch&#x3D;master
[build-url]: https://travis-ci.org/DarrenPaulWright/pathinator
[coverage]: https://coveralls.io/repos/github/DarrenPaulWright/pathinator/badge.svg?branch&#x3D;master
[coverage-url]: https://coveralls.io/github/DarrenPaulWright/pathinator?branch&#x3D;master
[deps]: https://david-dm.org/DarrenPaulWright/pathinator.svg
[deps-url]: https://david-dm.org/DarrenPaulWright/pathinator
[size]: https://packagephobia.now.sh/badge?p&#x3D;pathinator
[size-url]: https://packagephobia.now.sh/result?p&#x3D;pathinator
[vulnerabilities]: https://snyk.io/test/github/DarrenPaulWright/pathinator/badge.svg?targetFile&#x3D;package.json
[vulnerabilities-url]: https://snyk.io/test/github/DarrenPaulWright/pathinator?targetFile&#x3D;package.json
[license]: https://img.shields.io/github/license/DarrenPaulWright/pathinator.svg
[license-url]: https://npmjs.com/package/pathinator/LICENSE.md


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html)

## [0.4.1] - 2020-09-18
### Fixed
- Multiple consecutive shorthand lines now export correctly
- Better logic around exporting fractions when compress = true

## [0.4.0] - 2020-09-17
### Fixed
- Handling lack of whitespace around fractions and arc flags on import.
- Removing more whitespace around fractions and arc flags on export.
- Handling consecutive commands without letters.
- Some bugs around export scale and translate

### Added
- optimizeSvgPaths

## [0.3.3] - 2020-08-20
### Security
- Updating dependencies

## [0.3.2] - 2020-07-15
### Security
- Updating dependencies

## [0.3.1] - 2020-06-09
### Security
- Updating dependencies

## [0.3.0] - 2020-04-03
### Added
- path.transform

## [0.2.1] - 2020-04-03
### Added
- Command index parameter to path.eachPoint

## [0.2.0] - 2020-04-02
### Added
- Support for svg polygon elements
- path.eachPoint

## 0.1.0 - 2020-03-30
### Added
- Path

[0.4.0]: https://github.com/DarrenPaulWright/pathinator/compare/v0.3.3...v0.4.0
[0.3.3]: https://github.com/DarrenPaulWright/pathinator/compare/v0.3.2...v0.3.3
[0.3.2]: https://github.com/DarrenPaulWright/pathinator/compare/v0.3.1...v0.3.2
[0.3.1]: https://github.com/DarrenPaulWright/pathinator/compare/v0.3.0...v0.3.1
[0.3.0]: https://github.com/DarrenPaulWright/pathinator/compare/v0.2.1...v0.3.0
[0.2.1]: https://github.com/DarrenPaulWright/pathinator/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/DarrenPaulWright/pathinator/compare/v0.1.0...v0.2.0


src/Path.js
import { until as asyncUntil } from 'async-agent';
import { Schema } from 'hord';
import { Enum, isBoolean, isNumber, isString } from 'type-enforcer';
import { Point } from 'type-enforcer-math';
import Arc from './commands/Arc.js';
import Close from './commands/Close.js';
import Cubic from './commands/Cubic.js';
import Line from './commands/Line.js';
import Move from './commands/Move.js';
import Quadratic from './commands/Quadratic.js';
import origin from './utility/origin.js';

const commands = {
	m: (path, data) => path[add](Move, [data]),
	M: (path, data) => path[add](Move, [data, true]),
	l: (path, data) => path[add](Line, [data]),
	L: (path, data) => path[add](Line, [data, true]),
	h: (path, data) => path[add](Line, [data, 0], true),
	H: (path, data, currentPoint) => path[add](Line, [data, currentPoint.y, true], true),
	v: (path, data) => path[add](Line, [0, data], true),
	V: (path, data, currentPoint) => path[add](Line, [currentPoint.x, data, true], true),
	c: (path, data) => path[add](Cubic, [data]),
	C: (path, data) => path[add](Cubic, [data, true]),
	s: (path, data) => path[add](Cubic, [data], true),
	S: (path, data) => path[add](Cubic, [data, true], true),
	q: (path, data) => path[add](Quadratic, [data]),
	Q: (path, data) => path[add](Quadratic, [data, true]),
	t: (path, data) => path[add](Quadratic, [data], true),
	T: (path, data) => path[add](Quadratic, [data, true], true),
	a: (path, data) => path[add](Arc, [data]),
	A: (path, data) => path[add](Arc, [data, true]),
	z: (path) => path.close(),
	Z: (path) => path.close(true)
};

const coordinatesTo = {
	initial: (command, settings, nextCommands) => {
		settings.toAbsolute = command.isAbsolute();

		return command.export(settings, nextCommands);
	},
	absolute: (command, settings, nextCommands) => {
		settings.toAbsolute = true;

		return command.export(settings, nextCommands);
	},
	relative: (command, settings, nextCommands) => {
		settings.toAbsolute = false;

		return command.export(settings, nextCommands);
	},
	auto: (command, settings, nextCommands) => {
		const currentPoint = settings.currentPoint;
		const absolute = coordinatesTo.absolute(command, settings, nextCommands.slice());
		settings.currentPoint = currentPoint;
		settings.offset = undefined;
		const absoluteIsConsecutive = settings.isConsecutive;
		const relative = coordinatesTo.relative(command, settings, nextCommands);

		command.isExportedAbsolute = absolute.trim().length <= relative.trim().length;

		if (command.isExportedAbsolute) {
			settings.isConsecutive = absoluteIsConsecutive;
		}

		return command.isExportedAbsolute ? absolute : relative;
	}
};

const exportSettingsSchema = new Schema({
	coordinates: {
		type: Enum,
		enum: new Enum({
			initial: 'initial',
			absolute: 'absolute',
			relative: 'relative',
			auto: 'auto'
		})
	},
	compress: Boolean,
	combine: Boolean,
	fractionDigits: {
		type: 'integer',
		min: 0
	},
	scale: [Point, Object, Array, Number],
	translate: [Point, Object, Array, Number],
	maxCharsPerLine: {
		type: 'integer',
		min: 1
	},
	commandsOnNewLines: Boolean,
	toPolygon: Boolean,
	async: Boolean
});

const processTransformSettings = (settings) => {
	['scale', 'translate'].forEach((setting) => {
		const value = settings[setting];

		if (value !== undefined && !(value instanceof Point)) {
			if (isNumber(value)) {
				settings[setting] = new Point(value, value);
			}
			else {
				settings[setting] = new Point(value);
			}
		}
	});
};

const syncUntil = (callback) => new Promise((resolve) => {
	let result = false;

	while (!result) {
		result = callback();
	}

	resolve();
});

const add = Symbol();

const PATH = Symbol();
const END_OF_PATH = Symbol();
const SUB_PATH_START = Symbol();

/**
 * @typedef integer
 * @private
 */

/**
 * @name Installation
 * @summary
 *
 * ```
 * npm install pathinator
 * ```
 */

/**
 * Parse, build, and optimize SVG path data.
 *
 * @example
 * ``` javascript
 * import { Path } from 'pathinator';
 *
 * const path = new Path()
 *     .move(50, 100)
 *     .line(100, 100)
 *     .line(200, 200)
 *     .close();
 * ```
 *
 * @category 2
 * @class Path
 *
 * @param {string} [path] - Optional path data to parse.
 */
export default class Path {
	constructor(path) {
		this.import(path);
	}

	/**
	 * Import a path string. Removes any previous commands and create a new one.
	 *
	 * @memberOf Path
	 * @instance
	 * @chainable
	 *
	 * @param {string} path - A valid path data string or polygon string.
	 *
	 * @returns {object} Returns this.
	 */
	import(path) {
		this[PATH] = [];
		this[END_OF_PATH] = origin;
		this[SUB_PATH_START] = origin;

		if (isString(path)) {
			path = path.trim();

			const isPolygon = commands[path.charAt(0)] === undefined;
			let start = 0;

			if (isPolygon) {
				start = -1;
				let command = 'move';

				for (let index = 0; index <= path.length; index++) {
					if (path.charAt(index) === ' ' || index === path.length) {
						this[command](path.slice(start + 1, index), true);

						start = index;
						command = 'line';
					}
				}

				this.close(true);
			}
			else {
				for (let index = 1; index <= path.length; index++) {
					if (commands[path.charAt(index)] !== undefined || index === path.length) {
						commands[path.charAt(start)](this, path.slice(start + 1, index), this[END_OF_PATH]);

						start = index;
					}
				}
			}
		}

		return this;
	}

	[add](ThisCommand, args, isShorthand = false) {
		let isAbsolute = false;

		if (args !== undefined && isBoolean(args[args.length - 1])) {
			isAbsolute = args.pop();
		}

		ThisCommand.split(args, isShorthand)
			.forEach((arg) => {
				const previous = this[PATH][this[PATH].length - 1];
				const command = new ThisCommand(arg, previous, this[END_OF_PATH], isAbsolute);

				this[PATH].push(command);
				this[END_OF_PATH] = command.position(this[END_OF_PATH], this[SUB_PATH_START]);

				if (command instanceof Move) {
					this[SUB_PATH_START] = this[END_OF_PATH];
				}
			});

		return this;
	}

	/**
	 * Add a [move](https://www.w3.org/TR/SVG/paths.html#PathDataMovetoCommands) command.
	 *
	 * @memberOf Path
	 * @instance
	 * @chainable
	 *
	 * @param {...*} args - X and y coordinates or a string of X and y coordinates. If the final argument is `true` then command will be absolute coordinates.
	 *
	 * @returns {object} Returns this.
	 */
	move(...args) {
		return this[add](Move, args);
	}

	/**
	 * Add a [line](https://www.w3.org/TR/SVG/paths.html#PathDataLinetoCommands) command.
	 *
	 * @memberOf Path
	 * @instance
	 * @chainable
	 *
	 * @param {...*} args - X and y coordinates or a string of X and y coordinates. If the final argument is `true` then command will be absolute coordinates.
	 *
	 * @returns {object} Returns this.
	 */
	line(...args) {
		return this[add](Line, args);
	}

	/**
	 * Add a [quadratic bezier curve](https://www.w3.org/TR/SVG/paths.html#PathDataCubicBezierCommands) command.
	 *
	 * @memberOf Path
	 * @instance
	 * @chainable
	 *
	 * @param {...*} args - Series of coordinates or a string of coordinates. If the final argument is `true` then command will be absolute coordinates.
	 *
	 * @returns {object} Returns this.
	 */
	cubic(...args) {
		return this[add](Cubic, args);
	}

	/**
	 * Add a [quadratic bezier curve](https://www.w3.org/TR/SVG/paths.html#PathDataQuadraticBezierCommands) command.
	 *
	 * @memberOf Path
	 * @instance
	 * @chainable
	 *
	 * @param {...*} args - Series of coordinates or a string of coordinates. If the final argument is `true` then command will be absolute coordinates.
	 *
	 * @returns {object} Returns this.
	 */
	quadratic(...args) {
		return this[add](Quadratic, args);
	}

	/**
	 * Add an [arc](https://www.w3.org/TR/SVG/paths.html#PathDataEllipticalArcCommands) command.
	 *
	 * @memberOf Path
	 * @instance
	 * @chainable
	 *
	 * @param {...*} args - Series of coordinates / values or a string of coordinates / values. If the final argument is `true` then command will be absolute coordinates.
	 *
	 * @returns {object} Returns this.
	 */
	arc(...args) {
		return this[add](Arc, args);
	}

	/**
	 * Add a [close](https://www.w3.org/TR/SVG/paths.html#PathDataClosePathCommand) command.
	 *
	 * @memberOf Path
	 * @instance
	 * @chainable
	 *
	 * @param {boolean} [args=false] - If the argument is `true` then command will be absolute coordinates.
	 *
	 * @returns {object} Returns this.
	 */
	close(...args) {
		return this[add](Close, args);
	}

	/**
	 * Update command values at a specific index.
	 *
	 * @memberOf Path
	 * @instance
	 * @chainable
	 *
	 * @param {integer} index - Index of the command to update.
	 * @param {string|number[]} values - New values for the command at this index.
	 *
	 * @returns {object} Returns this.
	 */
	update(index, values) {
		const command = this[PATH][index];

		if (command !== undefined) {
			command.constructor.split([values])
				.forEach((value, newIndex) => {
					if (newIndex === 0) {
						command.set(value);
					}
				});
		}

		return this;
	}

	/**
	 * Calls  callback for each point in the path.
	 *
	 * @memberOf Path
	 * @instance
	 * @chainable
	 *
	 * @param {Function} callback - Provides three arguments: the Point, a boolean indicating if the point is a control point, and the command index.
	 *
	 * @returns {object} Returns this.
	 */
	eachPoint(callback) {
		this[PATH].forEach((command, index) => {
			command.eachPoint({
				currentPoint: origin,
				subPathStart: origin
			}, callback, index);
		});

		return this;
	}

	/**
	 * Transform all commands in path.
	 *
	 * @memberOf Path
	 * @instance
	 * @chainable
	 *
	 * @param {object} [settings] - Optional settings object.
	 * @param {integer} [settings.fractionDigits=3] - Round all numbers in path to a specified number of fraction digits.
	 * @param {number|Point|Array|object} [settings.scale] - Scale the entire path. If a number is provided then x and y are scaled the same. To scale x and y differently provide a Point, an array as [x, y], or an object as { x:_, y:_ }.
	 * @param {number|Point|Array|object} [settings.translate] - Translate the entire string a specified distance. If a number is provided then x and y are translated the same. To translated x and y differently provide a Point, an array as [x, y], or an object as { x:_, y:_ }.
	 *
	 * @returns {object} Returns this.
	 */
	transform(settings) {
		settings = {
			...settings,
			toAbsolute: true
		};

		processTransformSettings(settings);

		this[PATH].forEach((command) => {
			command.transform(settings);
		});

		return this;
	}

	/**
	 * Export a string of the path. Transforms are only applied to the exported path.
	 *
	 * @memberOf Path
	 * @instance
	 *
	 * @param {object} [settings] - Optional settings object.
	 * @param {string} [settings.coordinates=initial] - Can be 'absolute' to convert all coordinates to absolute, 'relative' to convert all coordinates to relative, 'auto' to convert coordinates to whichever is the fewest characters, 'initial' (default) to retain the coordinates set on each command.
	 * @param {boolean} [settings.compress=false] - Remove excess whitespace and unnecessary characters.
	 * @param {boolean} [settings.combine=true] - Combine consecutive commands that are redundant.
	 * @param {integer} [settings.fractionDigits=3] - Round all numbers in path to a specified number of fraction digits.
	 * @param {number|Point|Array|object} [settings.scale] - Scale the entire path. If a number is provided then x and y are scaled the same. To scale x and y differently provide a Point, an array as [x, y], or an object as { x:_, y:_ }.
	 * @param {number|Point|Array|object} [settings.translate] - Translate the entire string a specified distance. If a number is provided then x and y are translated the same. To translated x and y differently provide a Point, an array as [x, y], or an object as { x:_, y:_ }.
	 * @param {integer} [settings.maxCharsPerLine] - Add newlines at logical breaks in the path to improve readability.
	 * @param {boolean} [settings.commandsOnNewLines=false] - Add a newline between each command.
	 * @param {boolean} [settings.toPolygon=false] - Format the string for use in a polygon element. Sets coordinates to 'absolute'.
	 * @param {boolean} [settings.async=false] - Process each command in a separate Promise.
	 *
	 * @returns {Promise<string>}
	 */
	export(settings = {}) {
		const nextCommands = this[PATH].slice().reverse();
		const process = coordinatesTo[settings.coordinates] || coordinatesTo.initial;
		const until = settings.async === true ? asyncUntil : syncUntil;
		let output = '';

		exportSettingsSchema
			.validate(settings)
			.forEach((error) => {
				throw new TypeError(error.error, error);
			});

		settings = {
			coordinates: 'initial',
			combine: true,
			fractionDigits: 3,
			compress: false,
			toPolygon: false,
			...settings,
			currentPoint: origin,
			subPathStart: 0,
			previousNumber: ''
		};

		processTransformSettings(settings);

		if (settings.toPolygon) {
			settings.coordinates = 'absolute';
		}

		return until(() => {
			if (nextCommands.length !== 0) {
				const command = nextCommands.pop();
				const commandResult = process(command, settings, nextCommands);

				settings.offset = undefined;

				if (command instanceof Move) {
					settings.subPathStart = settings.currentPoint;
				}

				if (
					(settings.compress !== true &&
						commandResult !== '' &&
						output !== '' &&
						!settings.isConsecutive &&
						!settings.toPolygon &&
						output.charAt(output.length - 1) !== ' ') ||
					(settings.toPolygon && settings.compress === true)
				) {
					output += ' ';
				}

				output += commandResult;

				settings.previous = command;
				settings.isConsecutive = false;
			}

			return nextCommands.length === 0;
		})
			.then(() => output.trim());
	}

	/**
	 * The total number of commands in this path.
	 *
	 * @memberOf Path
	 * @instance
	 *
	 * @returns {integer}
	 */
	get length() {
		return this[PATH].length;
	}
}
